### PR TITLE
Add support for containers.conf

### PR DESCRIPTION
**Pull request by rhatdan**
This is a rework of Qi Wang's patches.
Import package pkg/config from containers/common to read containers.conf
This patch allows users to specify default values stored in containers.conf
that will modify the behaviour of buildah tool.

Signed-off-by: Qi Wang <qiwan@redhat.com>
Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

Replaces https://github.com/containers/buildah/pull/1953

### CODE REVIEW

File: pkg/cli/common.go [hunk 1]
```diff
@@ -220,6 +226,45 @@ func GetFromAndBudFlags(flags *FromAndBudResults, usernsResults *UserNSResults,
 	return fs
 }
 
+func SetDefaultConfig(cmd *cobra.Command, defaultConfig *config.Config) error {
```
Comment (QiWang19) [inline pkg/cli/common.go]: Do not need the function `SetDefaultConfig` because gloabl option `--containers-conf` was removed.

Comment (rhatdan) [reply pkg/cli/common.go]: Removed

File: pkg/cli/common.go [hunk 2]
```diff
@@ -279,3 +281,32 @@ func CheckAuthFile(authfile string) error {
 	}
 	return nil
 }
+
+// GetCapabilities returns the capabilities added to container
+func GetCapabilities(defaultCapabilities, addCapabilities, dropCapabilities []string) []string {
```
[the remark this answers is not in the document]
Comment (vrothberg) [reply pkg/cli/common.go]: Is this func a potential candidate for c/common ? I guess that's something libpod needs as well. Based on an earlier comment from @mheon in c/common I am on the lookout for code we could further consolidate in c/common.

File: tests/containers_conf.bats [hunk 3]
```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env bats
```
Comment (vrothberg) [inline tests/containers_conf.bats]: :100: for the tests!

File: tests/containers_conf.bats [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+@test "containers.conf env test" {
+    export CONTAINERS_CONF=${TESTSDIR}/containers.conf 
+    cid=$(buildah from --pull --signature-policy ${TESTSDIR}/policy.json docker.io/alpine)
+    run_buildah --log-level=error run $cid sh -c 'env | grep "foo=bar"'
+}
+
+@test "containers.conf selinux test" {
+    if ! which selinuxenabled > /dev/null 2> /dev/null ; then
+        skip "No selinuxenabled"
```
Comment (vrothberg) [inline tests/containers_conf.bats]: s/selinuxenabled/selinux enabled/

Comment (rhatdan) [reply tests/containers_conf.bats]: That is saying the selinuxenabled program does not exists.  So it should stay as a single name.  I will add "No selinuxenabled application"

File: cmd/buildah/common.go [hunk 5]
```diff
@@ -424,3 +428,13 @@ Flags:
 {{end}}
 `
 }
+
+func getDefaultConfig() *config.Config {
+	var err error
+	defaultConfig, err := config.NewConfig("")
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "%v\n", err)
+		os.Exit(1)
```
[the remark this answers is not in the document]
Comment (rhatdan) [reply cmd/buildah/common.go]: Fixed

File: cmd/buildah/common.go [hunk 6]
```diff
@@ -20,7 +21,10 @@ import (
 	"github.com/spf13/pflag"
 )
 
-var needToShutdownStore = false
+var (
+	defaultConfig       = getDefaultConfig()
```
[the remark this answers is not in the document]
Comment (rhatdan) [reply cmd/buildah/common.go]: Fixed

File: tests/containers.conf [hunk 7]
```diff
@@ -0,0 +1,368 @@
+# The containers configuration file specifies all of the available configuration
```
[the remark this answers is not in the document]
Comment (rhatdan) [reply tests/containers.conf]: Yes maybe we should remove all of the commented out ones.

**Comment by vrothberg** [comment]
"not ok 214 containers.conf additional devices test" see https://travis-ci.org/containers/buildah/jobs/634359720#L1031

**Comment by vrothberg** [comment]
# ls: /dev/fuse1: No such file or directory

# [ rc=1 (** EXPECTED 0 **) ]

# #/vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv

# #| FAIL: exit code is 1; expected 0

# #\^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

**Comment by rhatdan** [comment]
@vrothberg Now that I have this testing, can I bump containers/common version so we can merge this?

**Comment by vrothberg** [comment]
> @vrothberg Now that I have this testing, can I bump containers/common version so we can merge this?

@rhatdan as long as we strictly follow semantic versioning, sure. Any breaking change would mean a major version bump.

**Comment by rhatdan** [comment]
Since we are still a version 0, I don't believe this is true.  We are still working towards a stable release.

**Comment by vrothberg** [comment]
I reread through semver.org and you’re right. Everything below 1.0.0 is a
free ticket to do whatever we want 👌

**Comment by vrothberg** [comment]
Nice work, @QiWang19 and @rhatdan !

**Comment by TomSweeneyRedHat** [comment]
LGTM.
I ran into some issues testing the changes with one of the baseline tests, but have convinced myself it's not due to this change.  Most likely a system error.

**Comment by rhatdan** [comment]
@rh-atomic-bot r=TomSweeneyRedHat

**Pull request closed without merging**